fe_utils/solvers/helmholtz.py

```diff
@@ -28,6 +31,31 @@ def assemble(fs, f):
     l = np.zeros(fs.node_count)
 
     # Now loop over all the cells and assemble A and l
+    nodes = fs.cell_nodes
+    for c in range(nodes.shape[0]):
+        cell_nodes = fs.cell_nodes[c, :]
+        J = fs.mesh.jacobian(c)
+        detJ = np.abs(np.linalg.det(J))
+        l[cell_nodes] += (phi.T*Q.weights) @ (f.values[cell_nodes] @ phi.T) * detJ
+
+    for c in range(nodes.shape[0]):
+        J = fs.mesh.jacobian(c)
+        inv_J = np.linalg.inv(J)
+        detJ = np.abs(np.linalg.det(J))
+        # m is a ixj matrix for 1 2 it is 6x6
+        phi_i = Q.weights.T @ phi
+        phi_ij = phi_i.reshape(-1, 1) @ phi_i.reshape(1, -1)
+        phi_i_grad = np.tensordot(phi_grad, inv_J, axes=1)
+        phi_product = phi_i_grad @ phi_i_grad.swapaxes(1, 2)
+        A[np.ix_(nodes[c, :], nodes[c, :])] += \
+            (np.sum(Q.weights.reshape((-1,) + (1,)*(phi_product.ndim - 1))*phi_product, axis=0) + phi_ij)*detJ
```

Comment:
This is a really complicated way of writing this out. I am really not sure whether it is right or not. This code can be very simply implemented as a sum of two calls to `np.einsum` and it will be much easier to know whether you are write when written that way.